### PR TITLE
Show/change experiment default template in UI

### LINKED ISSUE #46: Add ui elements to show and change the default ui template of an experiment
The QHAna backend allows setting a default ui-template for each experiment. However, the QHAna UI currently does not expose this feature to the user.

The components should be added to the "Info"-tab of an experiment.

Relevant API routes in the backend:
- <https://github.com/UST-QuAntiL/qhana-backend/blob/349ff4c4466229b8e206ca39dd94027d42c371fd/main.bal#L1227>
- <https://github.com/UST-QuAntiL/qhana-backend/blob/349ff4c4466229b8e206ca39dd94027d42c371fd/main.bal#L1246>

Optionally extend the backend to include the template id (or the link to the API endpoints above) in the experiment response:
- <https://github.com/UST-QuAntiL/qhana-backend/blob/349ff4c4466229b8e206ca39dd94027d42c371fd/types.bal#L120>
- UI code that would use this: <https://github.com/UST-QuAntiL/qhana-ui/blob/db391ff4da92d023ec387a5ceb26ce7ad0dccd19/src/app/services/templates.service.ts#L62>

### PR DESCRIPTION
**Pull request by infacc**
Fixes #46

### CODE REVIEW

File: src/app/components/experiment/experiment.component.ts [hunk 1]
```diff
@@ -57,6 +64,15 @@ export class ExperimentComponent implements OnInit, OnDestroy {
             filter(value => value != null && value !== this.lastSavedDescription),
             debounceTime(500)
         ).subscribe(this.saveDescription);
+        this.registry.getByRel<PageApiObject>([["ui-template", "collection"]]).then(result => {
```
[the remark this answers is not in the document]
Comment (infacc) [reply src/app/components/experiment/experiment.component.ts]: The experiment default template selection is now paginated using the `mat-paginator` component:
image

File: src/app/components/experiment/experiment.component.html [hunk 2]
```diff
@@ -32,11 +32,21 @@ <h1 class="t-page-headline">Experiment</h1>
             </button>
         </mat-card-title>
         <mat-card-content>
+            <mat-form-field>
+                <mat-label>Default Template</mat-label>
+                <mat-select (selectionChange)="changeDefaultTemplate($event.value)" [value]="experimentTemplateId">
+                    <mat-option [value]="null">
+                        None
+                    </mat-option>
+                    <mat-option *ngFor="let template of experimentTemplates | keyvalue" [value]="template.key">
```
[the remark this answers is not in the document]
Comment (infacc) [reply src/app/components/experiment/experiment.component.html]: I refactored the `uiTemplates` property to a `Map`. However, when using `map.entries()` in the `*ngFor` directive the site throws an `ExpressionChangedAfterItHasBeenCheckedError`. Therefore, I still used the `keyvalue` pipe to iterate the `uiTemplates`. Should I further refactor the default template selection to work without the `keyvalue` pipe?

File: src/app/components/experiment/experiment.component.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (buehlefs) [reply src/app/components/experiment/experiment.component.ts]: A Paginator in a select dropdown works? This is OK I guess, but probably not great for acessability. Have you tried the growing list in there as well? The growing list would at least show the names of the templates by default. Just showing the numbers (also in the form field) is not good for usability. The dialog solution (while being more cumbersome to setup) is still my favorite for now. Maybe with a UI like this:

```
# test 🖉

Default template: [Template Name](#link-to-template-details-in-workspace?) [Button: Change Template] [Button: Remove Template]

┌───────────────────────────────────────────────────────────────────────────┐
│ Experiment description...
```

And the Dialog could look like this:

```
Select a default template:

<div style="max-height: something; overflow-y: auto">
    <growing-list />
</div>

Selected Template: {Template Name}

{Template Description}

--------------------------------
[Button: Unaset Template][Button: Set Default Template][Button: Cancel]
```

File: src/app/components/experiment/experiment.component.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (buehlefs) [reply src/app/components/experiment/experiment.component.html]: If angular requires the keyvalue pipe to be happy then use it. No need for further refactorings.

File: src/app/services/registry.service.ts [hunk 3]
```diff
@@ -483,7 +483,13 @@ export class PluginRegistryBaseService {
...
     public async getByApiLink<T>(link: ApiLink, searchParams: URLSearchParams | null = null, ignoreCache: boolean | "ignore-embedded" = false): Promise<ApiResponse<T> | null> {
         const url = new URL(link.href)
-        searchParams?.forEach((value, key) => url.searchParams.append(key, value));
+        searchParams?.forEach((value, key) => {
+            if (url.searchParams.has(key)) {
+                url.searchParams.set(key, value);
+            } else {
+                url.searchParams.append(key, value);
+            }
+        });
```
Comment (infacc) [inline src/app/services/registry.service.ts:492]: In the course of this pull request, I adjusted the `getByApiLink` method to replace existing parameters. The reason for this change was that setting the `item-count` URL parameter via the `searchParmas` for [`getByRel(['ui-template', 'collection'], searchParams)`](https://github.com/UST-QuAntiL/qhana-ui/blob/f6b830ca64462f9698eb7d269ecae19dedfe1e39/src/app/services/registry.service.ts#L459) didn't work. The `ApiLink` (given to `getByApiLink`) already contains the default parameter `item-count=25` and appending the parameter a second time doesn't work (Angular seems to use only the first occurrence).
As, currently, loading the templates is handled by a `qhana-growing-list` this change could be reverted. Does it make sense to revert this change?

Comment (buehlefs) [reply src/app/services/registry.service.ts:492]: No, keep your version. It is better if the extra url params actually are effective. However, in the future try to keep changes to such core/common parts in separate commits. This makes it easy to recognize these changes, and revert them later if necessary. It also allows you to write why that change was required in the first place into the commit message (not in the summary line, but the extended body). I use this to document smaller design decisions that are only ever relevant if one takes a look at the code.

File: src/app/components/experiment/experiment.component.ts [hunk 4]
```diff
@@ -172,12 +170,25 @@ export class ExperimentComponent implements OnInit, OnDestroy {
...
 
-    async changeDefaultTemplate(templateId: string | null) {
+    showSelectDefaultTemplateDialog() {
+        const dialogRef = this.dialog.open(ChooseTemplateComponent, {
+            minWidth: "20rem", maxWidth: "40rem", width: "60%", maxHeight: "95%",
+            data: this.uiTemplate
+        });
+        dialogRef.afterClosed().subscribe(templateId => {
+            if (templateId != null) {
+                const id = templateId === 'all-plugins' ? null : templateId;
```
Comment (buehlefs) [inline src/app/components/experiment/experiment.component.ts:180]: There is now a constant for the special `all-plugins` value (I believe in the templates service).

File: src/app/dialogs/choose-template/choose-template.component.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import { ApiLink } from 'src/app/services/api-data-types';
+import { PluginRegistryBaseService } from 'src/app/services/registry.service';
+import { TemplateApiObject } from 'src/app/services/templates.service';
+
+@Component({
+    selector: 'qhana-choose-template',
+    templateUrl: './choose-template.component.html',
+    styleUrls: ['./choose-template.component.sass']
+})
+export class ChooseTemplateComponent implements OnInit {
```
Comment (buehlefs) [inline src/app/dialogs/choose-template/choose-template.component.ts:12]: Dialogs should end with `Dialog` (not `Component`). (please rename everything, including the file names...)
The purpose of this is to more easily identify Dialogs and other UI Components as such (also Services, etc.)

File: src/app/services/registry.service.ts [hunk 6]
```diff
@@ -483,7 +483,13 @@ export class PluginRegistryBaseService {
 
     public async getByApiLink<T>(link: ApiLink, searchParams: URLSearchParams | null = null, ignoreCache: boolean | "ignore-embedded" = false): Promise<ApiResponse<T> | null> {
         const url = new URL(link.href)
-        searchParams?.forEach((value, key) => url.searchParams.append(key, value));
+        searchParams?.forEach((value, key) => {
+            if (url.searchParams.has(key)) {
+                url.searchParams.set(key, value);
+            } else {
+                url.searchParams.append(key, value);
```
Comment (buehlefs) [inline src/app/services/registry.service.ts:490]: What is the difference between `set` and `append`? Maybe we can just use `set` and skip the `has` test.

File: src/app/components/experiment/experiment.component.ts [hunk 7]
```diff
@@ -172,12 +170,25 @@ export class ExperimentComponent implements OnInit, OnDestroy {
...
+                this.updateExperimentDefaultTemplate(id);
+            }
+        });
+    }
+
+    async updateExperimentDefaultTemplate(templateId: string | null) {
         if (this.experimentId == null) {
+            console.warn("Experiment ID is null!");
```
Comment (buehlefs) [inline src/app/components/experiment/experiment.component.ts:188]: Should this show a warning?

File: src/app/components/experiment/experiment.component.html [hunk 8]
```diff
@@ -31,18 +31,21 @@ <h1 class="t-page-headline">Experiment</h1>
...
+                    Default Template:
+                </span>
+                <a [routerLink]="['/experiments', experimentId, 'workspace']" [queryParams]="{ template: uiTemplate?.self?.resourceKey?.uiTemplateId ?? 'all-plugins' }">
+                    {{uiTemplate?.name ?? 'All Plugins'}}
+                </a>
+                <button mat-icon-button (click)="showSelectDefaultTemplateDialog()">
+                    <mat-icon>edit</mat-icon>
+                </button>
+                <button mat-icon-button *ngIf="uiTemplate != null" (click)="updateExperimentDefaultTemplate(null)">
+                    <mat-icon>cancel</mat-icon>
```
Comment (buehlefs) [inline src/app/components/experiment/experiment.component.html:46]: provide a proper accesibility label text for these buttons (cancel is not what this button does)

**Review by buehlefs — CHANGES_REQUESTED** [review]

File: src/app/services/registry.service.ts (hunk 6 above)
Comment (infacc) [reply src/app/services/registry.service.ts:490]: You're right, we can just use `set`.

> The [set()](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/set) method of the URLSearchParams interface sets the value associated with a given search parameter to the given value. If there were several matching values, this method deletes the others. If the search parameter doesn't exist, this method creates it.

File: src/app/components/experiment/experiment.component.ts (hunk 7 above)
Comment (infacc) [reply src/app/components/experiment/experiment.component.ts:188]: This method is only used in the experiment info tab, where the `experimentId` should never be `null`. Does it make sense to show a warning in that case?

File: src/app/components/experiment/experiment.component.html (hunk 8 above)
Comment (infacc) [reply src/app/components/experiment/experiment.component.html:46]: The buttons are now labeled:

image

File: src/app/components/experiment/experiment.component.ts (hunk 7 above)
Comment (buehlefs) [reply src/app/components/experiment/experiment.component.ts:188]: It is also used in the html template by a button which calls this with `null` if I have read the code correctly.

Comment (buehlefs) [reply src/app/components/experiment/experiment.component.ts:188]: If you actually do not allow `null`, then remove it from the type of `templateId` and let the typescript type checker do the verification for you.

But then you have no way to unset the template id. Currently the unset Button does call this method with `templateId=null`: https://github.com/UST-QuAntiL/qhana-ui/blob/d917a689e43c2e52f7387d23ef03077ee403afc5/src/app/components/experiment/experiment.component.html#L45

**Review by buehlefs — APPROVED** [review]

File: src/app/dialogs/choose-template/choose-template.component.html
Comment (buehlefs) [inline src/app/dialogs/choose-template/choose-template.component.html:1]: Please also rename the html template and sass files to `*.dialog.*`

**Review by buehlefs — CHANGES_REQUESTED** [review]

File: src/app/dialogs/choose-template/choose-template.component.html
Comment (infacc) [reply src/app/dialogs/choose-template/choose-template.component.html:1]: Other dialogs also do not follow this convention. Should I rename them as well?

```
src/app/dialogs/
├── change-ui-template
│   ├── change-ui-template.component.html
│   ├── change-ui-template.component.sass
│   ├── change-ui-template.component.spec.ts
│   └── change-ui-template.component.ts
├── choose-data
│   ├── choose-data.component.html
│   ├── choose-data.component.sass
│   ├── choose-data.component.spec.ts
│   └── choose-data.component.ts
├── choose-plugin
│   ├── choose-plugin.component.html
│   ├── choose-plugin.component.sass
│   ├── choose-plugin.component.spec.ts
│   └── choose-plugin.component.ts
├── choose-template
│   ├── choose-template.dialog.html
│   ├── choose-template.dialog.sass
│   ├── choose-template.dialog.spec.ts
│   └── choose-template.dialog.ts
├── create-experiment
│   ├── create-experiment.component.html
│   ├── create-experiment.component.sass
│   ├── create-experiment.component.spec.ts
│   └── create-experiment.component.ts
├── delete-dialog
│   ├── delete-dialog.dialog.html
│   ├── delete-dialog.dialog.sass
│   └── delete-dialog.dialog.ts
├── export-experiment
│   ├── export-experiment.component.html
│   ├── export-experiment.component.sass
│   ├── export-experiment.component.spec.ts
│   └── export-experiment.component.ts
└── markdown-help
    ├── markdown-help.component.html
    ├── markdown-help.component.sass
    ├── markdown-help.component.spec.ts
    └── markdown-help.component.ts
```

Comment (buehlefs) [reply src/app/dialogs/choose-template/choose-template.component.html:1]: For this PR only rename the new dialog. If you want to rename the others (which should be done eventually) please open another PR.

**Pull request merged**